From 204b926dff4a529831fc194e4873c32f4719496a Mon Sep 17 00:00:00 2001
From: Stephen Boyd <swboyd@chromium.org>
Date: Wed, 10 Oct 2018 15:50:50 -0700
Subject: [PATCH] BACKPORT: Input: elants_i2c - use DMA safe i2c when possible

To avoid bounce buffer when an i2c controller decides to use DMA for a
transaction, let's make out buffer that we use for reads DMA-safe and let
the master know that DMAing into it is safe.

Signed-off-by: Stephen Boyd <swboyd@chromium.org>
Signed-off-by: Dmitry Torokhov <dmitry.torokhov@gmail.com>
(cherry picked from commit 00f73f97527f6ea8b4bc2bc63c44299bcc2f5cab)

Conflicts:
   drivers/input/touchscreen/elants_i2c.c

BUG=b:117520481
TEST=Built and ran on cheza

Change-Id: Id14a44d27e4d3584d9986560b707fa18b28bb9dd
Signed-off-by: Ryan Case <ryandcase@chromium.org>
Reviewed-on: https://chromium-review.googlesource.com/1308053
Reviewed-by: Stephen Boyd <swboyd@chromium.org>
---
 drivers/input/touchscreen/elants_i2c.c | 7 ++++---
 1 file changed, 4 insertions(+), 3 deletions(-)

diff --git a/drivers/input/touchscreen/elants_i2c.c b/drivers/input/touchscreen/elants_i2c.c
index b5c4783ea224..325ae6054d62 100644
--- a/drivers/input/touchscreen/elants_i2c.c
+++ b/drivers/input/touchscreen/elants_i2c.c
@@ -147,11 +147,12 @@ struct elants_data {
 	u8 cmd_resp[HEADER_SIZE];
 	struct completion cmd_done;
 
-	u8 buf[MAX_PACKET_SIZE];
-
 	bool wake_irq_enabled;
 	bool keep_power_in_suspend;
 	bool unbinding;
+
+	/* Must be last to be used for DMA operations */
+	u8 buf[MAX_PACKET_SIZE] ____cacheline_aligned;
 };
 
 static int elants_i2c_send(struct i2c_client *client,
@@ -864,7 +865,7 @@ static irqreturn_t elants_i2c_irq(int irq, void *_dev)
 	int i;
 	int len;
 
-	len = i2c_master_recv(client, ts->buf, sizeof(ts->buf));
+	len = i2c_master_recv_dmasafe(client, ts->buf, sizeof(ts->buf));
 	if (len < 0) {
 		dev_err(&client->dev, "%s: failed to read data: %d\n",
 			__func__, len);
-- 
2.23.0.rc1.153.gdeed80330f-goog

